File: dsl-api/src/main/java/cbs/dsl/api/EventTypes.java
package cbs.dsl.api;

import cbs.dsl.api.EventFunction.EventArg;
import cbs.dsl.api.EventFunction.EventResult;
import io.avaje.jsonb.Json;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.NoArgsConstructor;

import java.util.Map;

/** Consolidated event DSL types. */
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class EventTypes {

  @Json
  @Builder(toBuilder = true)
  public record EventInput(String eventNumber, Map<String, Object> params) implements EventArg {}

  @Json
  @Builder(toBuilder = true)
  public record EventOutput(Map<String, Object> params, EventStatus status) implements EventResult {

    public static EventOutput success(Map<String, Object> result) {
      return new EventOutput(result, EventStatus.SUCCESS);
    }
  }

  public enum EventStatus {
    UNDEFINED,
    SUCCESS,
    PENDING,
    ERROR,
  }
}

File: dsl-api/src/main/kotlin/cbs/dsl/api/EventTypes.kt
package cbs.dsl.api

import io.avaje.jsonb.Json

/**
 * Input carrier for event execution. Carries event parameters and execution metadata as
 * JSON-serializable data for Temporal workflow replay and PostgreSQL JSONB storage.
 */
@Json
data class EventInput(
    val params: Map<String, Any?>,
    val eventCode: String,
    val eventNumber: Long? = null,
    val workflowExecutionId: String? = null,
)

/**
 * Output carrier for event execution. Carries the enriched context and transaction results as
 * JSON-serializable data for Temporal workflow replay and PostgreSQL JSONB storage.
 */
@Json
data class EventOutput(
    val context: Map<String, Any?>,
    val transactionResults: Map<String, Map<String, Any?>>,
    val status: String = "SUCCESS",
)
